### PR TITLE
[CLIPPER-149] Container reconnection

### PR DESCRIPTION
**Pull request by Corey-Zumar**

### CODE REVIEW

**Comment by dcrankshaw** [comment]
@Corey-Zumar Should I review this yet?

**Comment by Corey-Zumar** [comment]
@dcrankshaw The code is fully functional. I just need to update the container implementation docs - will do that this evening and have the PR ready for review by tonight.

**Comment by Corey-Zumar** [comment]
@dcrankshaw This is ready for review

File: containers/container_implementation.md
Comment (dcrankshaw) [inline containers/container_implementation.md]: container is mis-spelled in the second sentence

File: containers/container_implementation.md
Comment (dcrankshaw) [inline containers/container_implementation.md]: Note whether the heartbeat is sent from Clipper to the container or the container to Clipper

File: containers/container_implementation.md
Comment (dcrankshaw) [inline containers/container_implementation.md]: Format the python code blocks in triple backticks like the following to get Python syntax highlighting
````
```py
 # Code goes here
```
````

File: containers/container_implementation.md
Comment (dcrankshaw) [inline containers/container_implementation.md]: Any objections to making the default timeout 30 seconds? I'd rather be slightly more aggressive with our defaults.

File: containers/container_implementation.md
Comment (dcrankshaw) [inline containers/container_implementation.md]: Very nice writeup. Welcome to the darkside of RPC implementations 😛. People (i.e. me) always start out by thinking that their RPC use case is simple and they should write their own. And then the implementation inevitably needs to get more complex as they realize they need things like heartbeating.

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java
Comment (dcrankshaw) [inline containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java]: Can we make these configurable at runtime?

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
I reviewed the revised protocol and things generally look good. I'll go over the code in more detail tomorrow. But we should definitely have an automated test for this. It doesn't have to be a unit test (unit testing this is likely to be hard), but can you write a script that we can use to test the RPC service implementation as part of continuous integration.

Probably the easiest way to do this is to create something like the RPC benchmark that uses just the RPC system on the Clipper side of things, and then create Python and Java testing containers. The script can start the containers, then run the RPC tester twice. The RPC tester can internally check the logic of the RPC system (correct messages sent, etc) and exit with a non-zero exit code if any problems are detected. By running the RPC tester twice (or just destroying the RPC system instance and creating it again in the C++ code), you can make sure that the reconnect mechanisms work. Let me know if you have questions about testing this.

File: containers/container_implementation.md
Comment (Corey-Zumar) [reply containers/container_implementation.md]: Fixed

File: containers/container_implementation.md
Comment (Corey-Zumar) [reply containers/container_implementation.md]: Done

File: containers/container_implementation.md
Comment (Corey-Zumar) [reply containers/container_implementation.md]: Done

File: containers/container_implementation.md
Comment (Corey-Zumar) [reply containers/container_implementation.md]: Sure! Done.

File: containers/container_implementation.md
Comment (Corey-Zumar) [reply containers/container_implementation.md]: Thank you! Heartbeating definitely made things more complicated (~1000 lines later...)

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java
Comment (Corey-Zumar) [reply containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java]: Runtime configuration of these parameters will create behavioral inconsistencies w.r.t RPC across container instances. I think all containers corresponding to a given language should maintain a consistent RPC protocol.

**Comment by Corey-Zumar** [comment]
@dcrankshaw 

Added a comprehensive test for correct RPC functionality. You can run the test by executing the `test_container_rpc.sh` script in `containers/tests`. The test operates as follows:

1. Containers maintain a timestamped log of recent RPC messages (stored via ringbuffer). 
2. The test script launches a Java container and a Python container, each with models having access to the RPC message log
3. The test script executes the `rpctest` target, starting Clipper's `rpc_container_test`
4. Clipper waits for the containers to connect. It then requests all recent log messages from Clipper based on the system time
5. Clipper verifies that the container's log messages are consistent with correct RPC protocol, logging any errors if they occur.
6. Clipper returns exit code `0` if the containers' protocols are valid, else it returns exit code `1`.
7. The test script repeats steps 3-6.
8. The test scripts exits with code `0` if both iterations of `rpc_container_test` were successful. Otherwise, it exits with code `1`.

`rpc_container_test` will fail automatically if containers are not validated within a configurable timeout (`test_container_rpc.sh` specifies a timeout of 20 seconds).

**Comment by dcrankshaw** [comment]
Reviewing now. Can you add your note about how the test works from the above Github comment as a comment at the top of `test_container_rpc.sh`?

File: containers/java/clipper-java-container/pom.xml
Comment (dcrankshaw) [inline containers/java/clipper-java-container/pom.xml:72]: This just packages all the containers into a single self-container jar so you don't have to worry about a bunch of classpath entries when running it

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java
Comment (dcrankshaw) [inline containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java:24]: What is RPCEventHistory?

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java
Comment (dcrankshaw) [inline containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java]: Sorry I forgot to remove commented out code. Remove this.

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java
Comment (dcrankshaw) [inline containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java]: Remove comment

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/logging/RPCEventHistory.java
Comment (dcrankshaw) [inline containers/java/clipper-java-container/src/main/java/clipper/container/app/logging/RPCEventHistory.java]: Is this just for logging?

File: containers/java/clipper-java-container/src/test/java/clipper/container/app/RPCEventHistoryTest.java
Comment (dcrankshaw) [inline containers/java/clipper-java-container/src/test/java/clipper/container/app/RPCEventHistoryTest.java:1]: Add maven tests to `./bin/run_unittests.sh`

Comment (dcrankshaw) [reply containers/java/clipper-java-container/src/test/java/clipper/container/app/RPCEventHistoryTest.java:1]: Also, delete `ContainerTest.java` because we already have an example container in `src/main` and it's not really a test. It will prevent `mvn test` from terminating.

File: containers/test/test_container_rpc.sh
Comment (dcrankshaw) [inline containers/test/test_container_rpc.sh:1]: Add a call to this script to `bin/run_unittests.sh`. Now we can test the RPC service in CI. (Don't be worried if it fails on Jenkins at first. The jenkins machines may be missing some dependencies that we need to install.

File: containers/test/test_container_rpc.sh
Comment (dcrankshaw) [inline containers/test/test_container_rpc.sh]: This script doesn't start Redis. That's fine, because it should be called from `run_unittests.sh` which does start Redis. But it needs to take the Redis port as an argument because that script starts Redis on a random port.

Modify the calls to `rpctest` to the following:

```sh
# $1 is the first argument to the script
REDIS_PORT=$1
./rpctest --num_containers=2 --timeout_seconds=20 --redis_port $REDIS_PORT
redis-cli -p $REDIS_PORT "flushall"
echo "Sleeping for 5 seconds..."
sleep 5s
echo "Executing RPC test (second iteration)..."
./rpctest --num_containers=2 --timeout_seconds=20 --redis_port $REDIS_PORT
redis-cli -p $REDIS_PORT "flushall"
```

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
Overall this looks good. No major changes. Given the size of the PR there might be things that got lost in the details, but I didn't see anything.

Address my comments then I'm going to do some more testing (run the test again, make sure unit tests still pass with this new addition, make sure we didn't break the tutorial, etc). But I think we should be able to get this merged soon.

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java
Comment (Corey-Zumar) [reply containers/java/clipper-java-container/src/main/java/clipper/container/app/ModelContainer.java:24]: Added class documentation. This is a circular buffer containing the most recent RPC events involving the container.

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java
Comment (Corey-Zumar) [reply containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java]: Done

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java
Comment (Corey-Zumar) [reply containers/java/clipper-java-container/src/main/java/clipper/container/app/RPCProtocolTest.java]: Done

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/logging/RPCEventHistory.java
Comment (Corey-Zumar) [reply containers/java/clipper-java-container/src/main/java/clipper/container/app/logging/RPCEventHistory.java]: In the sense that it's for maintaining a log of recent RPC events to be used in validation of the RPC protocol, yes. Open to different package name suggestions if something else is more descriptive/accurate

File: containers/java/clipper-java-container/src/test/java/clipper/container/app/RPCEventHistoryTest.java
Comment (Corey-Zumar) [reply containers/java/clipper-java-container/src/test/java/clipper/container/app/RPCEventHistoryTest.java:1]: Done! Also fixed the tests...

File: containers/test/test_container_rpc.sh
Comment (Corey-Zumar) [reply containers/test/test_container_rpc.sh:1]: Done

File: containers/test/test_container_rpc.sh
Comment (Corey-Zumar) [reply containers/test/test_container_rpc.sh]: Done! Thanks for the code

**Comment by Corey-Zumar** [comment]
@dcrankshaw Addressed your comments. You were right about jenkins missing some dependencies.

File: containers/java/clipper-java-container/src/main/java/clipper/container/app/logging/RPCEventHistory.java
Comment (dcrankshaw) [reply containers/java/clipper-java-container/src/main/java/clipper/container/app/logging/RPCEventHistory.java]: No that's fine. I just wasn't sure what the purpose of the event history was (and the RPC protocol test was one of the last files I looked at when reviewing, so I saw this file first).

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
ok, we're using the default system java8...  build still failing, but i've at least got that bit sorted.

**Comment by shaneknapp** [comment]
test this please

**Comment by dcrankshaw** [comment]
I think Jenkins should be able to run the tests now.

@Corey-Zumar can you update to the latest on develop and I'll do a last round of reviews before merging?

**Comment by Corey-Zumar** [comment]
@dcrankshaw Rebased on develop

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
Boom! Tests passed.

**Review by dcrankshaw — APPROVED** [review]

**Comment by shaneknapp** [comment]
nice!

**Comment by shaneknapp** [comment]
for reference:
1) i had to remove the jenkins-managed JDK as it would override the default system JDK (7 vs 8)
2) built jmzq and disted it to the workers w/ansible:  https://github.com/ucbrise/jenkins-infra/pull/49
3) profit

**Pull request merged**
